File: java/account/DateDelegate.java
/**
 * @file AccountDelegate.java
 * @author Ronald T. Fernandez
 * @version 1.0
 */
package account;

import java.util.Calendar;

import corbaAccount.date;

import connection.Connection;

/**
 * This class implements the delegate for date.
 * @see DateImpl
 */
public class DateDelegate {

	/**
	 * This is an instance of an DateImpl, in which the class delegates
	 */
	private DateImpl _instance;
	
	/**
	 * Class constructor
	 * @throws Exception Throws an exception in case of any problem
	 */
	public DateDelegate() throws Exception {
		this(true);
	}

	/**
	 * Class constructor
	 * @param publish	Indicates whether the date needs to be added to the corba naming service
	 * @throws Exception Throws an exception in case of any problem
	 */
	public DateDelegate(boolean publish) throws Exception {
		_instance = new DateImpl();
		if (publish) {
			Calendar c = Calendar.getInstance();
			// We assume that two instances of date can be created on the same second. Otherwise, we could skip the milliseconds.
			String time = String.format("%02d", Integer.valueOf(c.get(Calendar.HOUR))) + ":" + 
							String.format("%02d", Integer.valueOf(c.get(Calendar.MINUTE))) + ":" + 
							String.format("%02d", Integer.valueOf(c.get(Calendar.SECOND))) +
							String.format("%02d", Integer.valueOf(c.get(Calendar.MILLISECOND)));
			// Appends the current time to the reference name in CORBA the naming service
			Connection.getInstance().bindObjectToName(_instance._this(), "myContext", "Date"+_instance._toString() + "_" + time, "Date");
		}
	}	
	
	/**
	 * Class constructor
	 * @param year	Year
	 * @param month	Month
	 * @param day	Day
	 * @throws Exception Throws an exception in case of any problem
	 */
	public DateDelegate(int year, int month, int day) throws Exception {
		this(year, month, day, true);
	}

	/**
	 * @param year	Year
	 * @param month	Month
	 * @param day	Day
	 * @param publish	Indicates whether the date needs to be added to the corba naming service
	 * @throws Exception Throws an exception in case of any problem
	 */
	public DateDelegate(int year, int month, int day, boolean publish) throws Exception {
		_instance = new DateImpl(year,  month,  day);
		if (publish)
			Connection.getInstance().bindObjectToName(_instance._this(), "myContext", "Date"+_instance._toString(), "Date");
	}
	
	/**
	 * Get the date in String format
	 * @return Date in String format
	 */
	public String toString() {
		return _instance._toString();
	}
	
	/**
	 * Get the year
	 * @return Year
	 */
	public int getYear() {
		return _instance.year();
	}

	/**
	 * Get the month
	 * @return Month
	 */
	public int getMonth() {
		return _instance.month();
	}
	
	/**
	 * Get the day
	 * @return Day
	 */
	public int getDay() {
		return _instance.day();
	}
	
	/**
	 * Get the CORBA instance of the date
	 * @return date CORBA instance
	 */
	public date getCorbaInstance() {
		return _instance._this();
	}
	
}

File: cxx/account/DateDelegate.cpp
/**
 * \file DateDelegate.cpp
 * \author Ronald T. Fernandez
 * \mail ronaldtfernandez@gmail.com
 * \version 1.0
 */

#include "DateDelegate.h"

#include "../utils/Utils.h"

#include <chrono>
#include <ctime>

namespace account {

DateDelegate::DateDelegate(bool publish) {
	_instance = std::unique_ptr<DateImpl>(new DateImpl);
	if (publish) {
		std::string time;
		std::chrono::system_clock::time_point timepoint = std::chrono::system_clock::now();
		time_t t = std::chrono::system_clock::to_time_t(timepoint);
		connection::Connection::getInstance()->bindObjectToName(_instance->_this(), "myContext", std::string("Date") + std::string(_instance->toString()) + "_" + utils::Utils::convertDate(t), "Date");
	}

}

DateDelegate::DateDelegate(const DateDelegate& dd) {
	_instance = std::unique_ptr<DateImpl>(new DateImpl(dd.getYear(), dd.getMonth(), dd.getDay()));
}

DateDelegate::~DateDelegate() {
}

DateDelegate::DateDelegate(corbaAccount::date_ptr d) {
	_instance = std::unique_ptr<DateImpl>(new DateImpl(d->year(), d->month(), d->day()));
}

DateDelegate::DateDelegate(int year, int month, int day, bool publish) {
	_instance = std::unique_ptr<DateImpl>(new DateImpl(year,  month,  day));
	if (publish)
		connection::Connection::getInstance()->bindObjectToName(_instance->_this(), "myContext", "Date"+ std::string(_instance->toString()), "Date");
}

std::string DateDelegate::toString() const {
	return _instance->toString();
}

int DateDelegate::getYear() const {
	return _instance->year();
}

int DateDelegate::getMonth() const {
	return _instance->month();
}

int DateDelegate::getDay() const {
	return _instance->day();
}

corbaAccount::date_ptr DateDelegate::getCorbaInstance() const {
	return _instance->_this();
}

} /* namespace account */
